File: content/snippets/python/interview_lcci_01_06_compress_string.py
def compress_string(s: str) -> str:
    if not s:
        return s

    out: list[str] = []
    run = 1

    for i in range(1, len(s) + 1):
        if i < len(s) and s[i] == s[i - 1]:
            run += 1
            continue

        out.append(s[i - 1])
        out.append(str(run))
        run = 1

    compressed = "".join(out)
    return compressed if len(compressed) < len(s) else s


def main() -> None:
    print(f"aabcccccaaa -> {compress_string('aabcccccaaa')}")
    print(f"abbccd -> {compress_string('abbccd')}")


if __name__ == "__main__":
    main()

File: content/snippets/go/interview_lcci_01_06_compress_string.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func compressString(s string) string {
	if len(s) == 0 {
		return s
	}

	var b strings.Builder
	run := 1

	for i := 1; i <= len(s); i++ {
		if i < len(s) && s[i] == s[i-1] {
			run++
			continue
		}

		b.WriteByte(s[i-1])
		b.WriteString(strconv.Itoa(run))
		run = 1
	}

	compressed := b.String()
	if len(compressed) < len(s) {
		return compressed
	}
	return s
}

func main() {
	fmt.Printf("aabcccccaaa -> %s\n", compressString("aabcccccaaa"))
	fmt.Printf("abbccd -> %s\n", compressString("abbccd"))
}
